desc/objectives/_free_boundary.py

```diff
@@ -0,0 +1,840 @@
+"""Objectives for solving free boundary equilibria."""
+
+import warnings
+
+import numpy as np
+from scipy.constants import mu_0
+
+from desc.backend import jnp
+from desc.compute import compute as compute_fun
+from desc.compute import get_params, get_profiles, get_transforms, xyz2rpz_vec
+from desc.grid import LinearGrid
+from desc.nestor import Nestor
+from desc.objectives.objective_funs import _Objective
+from desc.singularities import (
+    DFTInterpolator,
+    FFTInterpolator,
+    virtual_casing_biot_savart,
+)
+from desc.utils import Timer, errorif, warnif
+
+from .normalization import compute_scaling_factors
+
+
+class VacuumBoundaryError(_Objective):
+    """Target B*n = 0 and B^2_in - B^2_out = 0 on LCFS.
+
+    Parameters
+    ----------
+    eq : Equilibrium
+        Equilibrium that will be optimized to satisfy the Objective.
+    ext_field : MagneticField
+        External field produced by coils.
+    target : float, ndarray, optional
+        Target value(s) of the objective. Only used if bounds is None.
+        len(target) must be equal to Objective.dim_f
+    bounds : tuple, optional
+        Lower and upper bounds on the objective. Overrides target.
+        len(bounds[0]) and len(bounds[1]) must be equal to Objective.dim_f
+    weight : float, ndarray, optional
+        Weighting to apply to the Objective, relative to other Objectives.
+        len(weight) must be equal to Objective.dim_f
+    normalize : bool
+        Whether to compute the error in physical units or non-dimensionalize.
+    normalize_target : bool
+        Whether target and bounds should be normalized before comparing to computed
+        values. If `normalize` is `True` and the target is in physical units,
+        this should also be set to True.
+    grid : Grid, optional
+        Collocation grid containing the nodes to evaluate error at.
+    field_grid : Grid, optional
+        Grid used to discretize ext_field.
+    name : str
+        Name of the objective function.
+
+    """
+
+    _scalar = False
+    _linear = False
+    _print_value_fmt = "Boundary Error: {:10.3e} "
+    _units = "(T)"
+    _coordinates = "rtz"
+
+    def __init__(
+        self,
+        eq,
+        ext_field,
+        target=None,
+        bounds=None,
+        weight=1,
+        normalize=True,
+        normalize_target=True,
+        grid=None,
+        field_grid=None,
+        name="Vacuum boundary error",
+    ):
+        if target is None and bounds is None:
+            target = 0
+        self._grid = grid
+        self._ext_field = ext_field
+        self._field_grid = field_grid
+        things = [eq]
```

Comment:
If we were to have  single stage, where the external field is also allowed to vary, would we have a separate quadratic flux objective with the external field having DOF for it, or would we allow it to be inside of `things` for the free boundary objectives? I guess the former probably

Reply:
yeah i need to add field here

Reply:
I guess we can leave that to a future PR though? I will make an issue just to keep a tab on this #855